File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/RemoveDuplicatesfromSortedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        if head == None:
            return
        curr = head
        while curr != None:
            next_node = curr.next
            while next_node != None and next_node.val == curr.val:
                next_node = next_node.next
            curr.next = next_node
            curr = next_node
        return head

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        if head == None:
            return
        next_node = head.next
        while next_node != None and next_node.val == head.val:
            next_node = next_node.next
        head.next = next_node
        self.deleteDuplicates(next_node)
        return head

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Ruby/remove_duplicates_from_sorted_list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  if head.nil? || head.next.nil?
    return head
  end

  prev_val = head.val
  prev, curr = head, head.next

  while curr
    if curr.val == prev_val
      # remove curr_node
      prev.next = curr.next
      curr = curr.next
    else
      prev_val = curr.val
      prev, curr = curr, curr.next
    end
  end
  return head
end
